File: examples/algorithms-cpp/ac_automaton.cpp
/**
 * ac_automaton.cpp - C++ AC自动机实现
 * 多模式字符串匹配
 */

#include <map>
#include <string>
#include <vector>
#include <queue>
#include <memory>

namespace ac {

struct Node {
    std::map<char, std::shared_ptr<Node>> children;
    std::shared_ptr<Node> fail;
    std::vector<std::string> output;
};

class AhoCorasick {
    std::shared_ptr<Node> root;
public:
    AhoCorasick() : root(std::make_shared<Node>()) {}

    void insert(const std::string& word) {
        auto node = root;
        for (char ch : word) {
            if (!node->children.count(ch)) {
                node->children[ch] = std::make_shared<Node>();
            }
            node = node->children[ch];
        }
        node->output.push_back(word);
    }

    void build() {
        std::queue<std::shared_ptr<Node>> q;
        for (auto& pair : root->children) {
            pair.second->fail = root;
            q.push(pair.second);
        }
        while (!q.empty()) {
            auto curr = q.front(); q.pop();
            for (auto& pair : curr->children) {
                char ch = pair.first;
                auto child = pair.second;
                auto fail = curr->fail;
                while (fail != nullptr && !fail->children.count(ch)) {
                    fail = fail->fail;
                }
                child->fail = (fail != nullptr) ? fail->children[ch] : root;
                if (child->fail == nullptr) child->fail = root;
                child->output.insert(child->output.end(), child->fail->output.begin(), child->fail->output.end());
                q.push(child);
            }
        }
    }

    std::vector<std::pair<int, std::vector<std::string>>> search(const std::string& text) {
        std::vector<std::pair<int, std::vector<std::string>>> res;
        auto node = root;
        for (size_t i = 0; i < text.size(); i++) {
            char ch = text[i];
            while (node != root && !node->children.count(ch)) {
                node = node->fail;
            }
            node = node->children.count(ch) ? node->children[ch] : root;
            if (!node->output.empty()) {
                res.push_back({(int)i, node->output});
            }
        }
        return res;
    }
};

} // namespace ac

File: examples/algorithms-java/ACAutomaton.java
/**
 * Aho-Corasick 自动机
 * 多模式字符串匹配
 * 构建: O(总模式长度), 查询: O(文本长度 + 匹配数)
 */
import java.util.*;

public class ACAutomaton {
    static class Node {
        Map<Character, Node> children = new HashMap<>();
        Node fail;
        List<String> output = new ArrayList<>();
    }

    private final Node root = new Node();

    public void insert(String word) {
        Node node = root;
        for (char ch : word.toCharArray()) {
            node = node.children.computeIfAbsent(ch, k -> new Node());
        }
        node.output.add(word);
    }

    public void build() {
        Queue<Node> q = new LinkedList<>();
        for (Node child : root.children.values()) {
            child.fail = root;
            q.add(child);
        }
        while (!q.isEmpty()) {
            Node curr = q.poll();
            for (Map.Entry<Character, Node> entry : curr.children.entrySet()) {
                char ch = entry.getKey();
                Node child = entry.getValue();
                Node fail = curr.fail;
                while (fail != null && !fail.children.containsKey(ch)) {
                    fail = fail.fail;
                }
                child.fail = (fail != null) ? fail.children.getOrDefault(ch, root) : root;
                child.output.addAll(child.fail.output);
                q.add(child);
            }
        }
    }

    public List<Map.Entry<Integer, List<String>>> search(String text) {
        List<Map.Entry<Integer, List<String>>> res = new ArrayList<>();
        Node node = root;
        for (int i = 0; i < text.length(); i++) {
            char ch = text.charAt(i);
            while (node != root && !node.children.containsKey(ch)) {
                node = node.fail;
            }
            node = node.children.getOrDefault(ch, root);
            if (!node.output.isEmpty()) {
                res.add(new AbstractMap.SimpleEntry<>(i, new ArrayList<>(node.output)));
            }
        }
        return res;
    }
}
